common/thorhelper/thorsoapcall.cpp

```diff
@@ -389,27 +389,47 @@ class BlackLister : public CInterface, implements IThreadFactory
             throw MakeStringException(-1, "blacklisted socket %s", s.str());
         }
         Owned<IException> exc;
-        try
+
+        unsigned numAttemptsRemaining = retries+1;
+        if (!numAttemptsRemaining)
+            numAttemptsRemaining++;
+
+        unsigned connectTimeMS = timeoutMS;
+        if (timeoutMS == WAIT_FOREVER)
+            connectTimeMS = 60000 / numAttemptsRemaining;
+
+        while(numAttemptsRemaining-- > 0)
         {
-            checkRoxieAbortMonitor(roxieAbortMonitor);
-            Owned<ISocket> sock;
-            Owned<ISocketConnectWait> scw = nonBlockingConnect(ep, timeoutMS == WAIT_FOREVER ? 60000 : timeoutMS*(retries+1));
-            for (;;)
+            try
             {
-                sock.setown(scw->wait(1000));//throws if connect fails or timeoutMS
                 checkRoxieAbortMonitor(roxieAbortMonitor);
-                if (sock)
-                    return sock.getLink();
+                Owned<ISocket> sock;
+                Owned<ISocketConnectWait> scw = nonBlockingConnect(ep, connectTimeMS);
+                for (;;)
+                {
+                    sock.setown(scw->wait(1000)); //throws if connect fails or timeoutMS
+                    checkRoxieAbortMonitor(roxieAbortMonitor);
+                    if (sock)
+                        return sock.getLink();
+                }
+            }
+            catch (IJSOCK_Exception *e)
+            {
```

Comment:
Probably also need the following:

```
if (e->errorCode()) == ROXIE_ABORT_EVENT)
    throw;

```

Reply:
yes, thanks.

Reply:
Maybe it didn't need that because it was catching a socket exception.  I'm not sure if I prefer new code or not.

Reply:
Sure, either way ok with me.